[ios] Red for over speed; Speed limit for Carplay

AntonM030481 opened the pull request:
### Summary
iPhone: 
- Red speed digits for overspeed.

CarPlay: 
- Show speed limit. 
- Red speed digits for overspeed. 
- Blinking for speed cam.

### Coding
Continuation of refactoring to unify all speed conversions and related localizations.
Use MeterPerSecond everywhere. Use numeric speed instead of string.
Rename of SpeedLimit to SpeedCamLimit to not confuse with usual limits (not speed cam).

### iPhone

Unknown limit
<img src="https://user-images.githubusercontent.com/85622715/181247153-80082ac8-9e98-486d-8e70-009676ef58a5.png" width="300">

Known limit and no over speed
<img src="https://user-images.githubusercontent.com/85622715/181247711-2cf223de-4975-4d88-bc4a-d79e46e8a8c6.png" width="300">

Known limit and over speed
<img src="https://user-images.githubusercontent.com/85622715/181236321-87ee0689-cf6e-4f14-adc7-05a5cce2136f.png" width="300">

Unknown speed cam (more exactly - unknown road speed limit) and over speed
<img src="https://user-images.githubusercontent.com/85622715/181261029-64c8f3dc-75aa-461f-81bd-6547a0f45110.png" width="300">

Known speed cam (more exactly - known road speed limit) and over speed
<img src="https://user-images.githubusercontent.com/85622715/181258964-de09ceba-8a8e-432a-b019-3214a7721178.png" width="300">

### CarPlay

Limit unknown
<img src="https://user-images.githubusercontent.com/85622715/181245279-f2a5554b-943c-41e4-a606-995fd20d5074.png" width="500">

Known limit and no over speed (speed sign is visible)
<img src="https://user-images.githubusercontent.com/85622715/181707248-7f84223b-3f62-4985-90a3-a3bb7af4da5d.png" width="500">

Known limit and over speed (speed sign is visible; current speed text is red)
<img src="https://user-images.githubusercontent.com/85622715/181708205-14bbb668-5d08-4127-99d7-e9d9fffe4fd9.png" width="500">

Unknown SpeedCam detected (red speed cam icon is visible)
<img src="https://user-images.githubusercontent.com/85622715/181238956-4b9c65ce-c9fb-4738-8063-14d4712a5d14.png" width="500">

Known SpeedCam detected and no over speed (speed sign icon is visible is switching with speed cam icon; current speed background is green)
<img src="https://media.giphy.com/media/ZTsbeEIbRyvlEF0kpV/giphy.gif" width="500">

Known SpeedCam detected and over speed (speed sign icon is visible is switching with speed cam icon; current speed background is red)
<img src="https://media.giphy.com/media/tQjeilrIvEXCQ6I7x8/giphy.gif" width="500">

biodranik reviewed (changes requested):
Отредактируй коммит через amend и удали модификации сабмодулей, пожалуйста. Откуда они взялись?

biodranik commented on platform/localization.hpp:
```diff
@@ -2,6 +2,11 @@
 
+namespace measurement_utils
+{
+  enum class Units;
```
Отступ не нужен.

AntonM030481 replied:
OK

biodranik commented on platform/measurement_utils.cpp:
```diff
@@ -224,14 +224,19 @@ string FormatSpeed(double metersPerSecond)
+
+string FormatSpeedNumeric(double metersPerSecond, Units units)
+{
+  double unitsPerHour = MpsToUnits(metersPerSecond, units);
```
const

AntonM030481 replied:
OK

biodranik commented on routing/following_info.hpp:
```diff
@@ -84,9 +84,8 @@ class FollowingInfo
-  std::string m_speedLimitUnitsSuffix;
+  // Current speed limit in meters per second.
+  // If no info about speed limit then m_speedLimitMps < 0.
+  double m_speedLimitMps = -1.0;
```
Лучше всё-таки везде длиннее, но однозначнее, тогда комментарий не нужен.
```suggestion
  double m_speedLimitMetersPerSecond = -1.0;
```

vng replied:
Мой имхо MetersPerSecond  - перебор. С чем это можно перепутать?

biodranik replied:
Как с чем? Я читаю всегда сначала как miles per second, стандартная аббревиатура MPS.

biodranik replied:
И многие американцы/англичане тоже могут так читать.

vng replied:
Саша, miles per second? Та на ракете летаешь? Нет такой стандартной аббревиатуры :)

AntonM030481 replied:
Мили - это всегда Mi,
MPH - единственное исключение.

biodranik replied:
Столько спора из-за _более понятной_ строчки ) Или мой вариант _менее понятен_?

AntonM030481 replied:
Тогда везде надо исправлять. Необходимо единообразие как раз во избежание недопонимания.
У нас скорость как правило именно в метрах в секунду и почти везде постфикс mps.

vng replied:
Лично у меня такие длинные название переменных m_speedLimitMetersPerSecond вызывают диссонанс.

biodranik commented on routing/routing_session.cpp:
```diff
@@ -417,7 +407,9 @@ void RoutingSession::GetRouteFollowingInfo(FollowingInfo & info) const
-    FormatSpeed(speedLimit.GetSpeedKmPH(), info.m_speedLimit, info.m_speedLimitUnitsSuffix);
+    info.m_speedLimitMps = measurement_utils::KmphToMps(speedLimit.GetSpeedKmPH());
+  else
+    info.m_speedLimitMps = -1.0;
```
Почему именно отрицательное значение выбрано, а не ноль, например? Если это особая константа, то лучше её вынести явно.

AntonM030481 replied:
А потом сравнивать с этой константой? Насколько это уместно для double?

biodranik replied:
Т.е. критерий выбора — невалидность определяется как < 0? Тогда всё равно константа делает код нагляднее.

На ноль даблы можно смело сравнивать, если они не модифицируюся и не являются результатом вычисления, а только следствием явного присваивания. У нас включен fast math если что, там вообще многие проверки на даблы убраны.

AntonM030481 replied:
В связи с тем что сейчас CarPlay c iPhone никак не унифицирован, не очень удобно пользоваться константами.
Оставил коменты, потом унифицирую и сделаю константами.

biodranik reviewed (changes requested):
@vng посмотри, пожалуйста, помержим и я потестирую.

@AntonM030481 дай знать, если нужна помощь с гитом, редактированием коммитов и ПРов и выпиливанием сабмодулей. Не нужно закрывать ПРы просто так )

AntonM030481 commented:
@biodranik хорошо)
Я просто все заново красиво форс репушил, и PR закрылся в процессе (откатился на самое начало). А кнопку Reopen я не заметил.

vng commented on iphone/Maps/UI/CarPlay/CarPlayMapViewController.swift:
```diff
@@ -38,73 +40,117 @@ final class CarPlayMapViewController: MWMViewController {
+      speedCamLimitContainer.layer.borderColor = UIColor.speedLimitRed().cgColor
+      speedCamImageView.tintColor = UIColor.speedLimitRed()
+
+      if let speedCamLimitMps = (speedCamLimitMps ?? speedLimitMps) {
```
Это какая-то фишка swift когда speedCamLimitMps сразу и слева и справа от присваивания?
Может тут self. добавить для наглядности?

AntonM030481 replied:
Conditional unwrapping. Чаще всего используется с тем же именем.
Прояснил в другом месте.

biodranik replied:
Я не понимаю, зачем отдельно speedcam limit. Когда он отличается от обычного speed limit? Зачем усложнять?

AntonM030481 replied:
speedCamLimitMps - это лимит именно от камеры.
speedLimitMps - это лимит дороги.
Иногда у камеры проставлено значение, иногда у дороги.
Скорость камеры имеет приоритет.
В общем случае лимит скорость камеры может не совпадать с текущим лимитом дороги,
потому что мы предупреждаем о камере заранее и, возможно, там где будет камера, ограничение дороги будет отличаться от текущего ограничения дороги.

AntonM030481 replied:
добавил self

AntonM030481 replied:
добавил комментарии про [это](https://github.com/organicmaps/organicmaps/pull/3047#discussion_r935639978).

vng commented on iphone/Maps/Classes/CustomViews/NavigationDashboard/Views/NavigationControlView.swift:
```diff
@@ -155,23 +155,34 @@ final class NavigationControlView: SolidTouchView, MWMTextToSpeechObserver, MapO
+
     speedLabel.text = speed
-    speedLegendLabel.text = info.speedUnits
+    speedLegendLabel.text = speedMeasure.unit
```
Да, просто C++ -> ObjC функция для использования в Swift.

vng replied:
На крайний случай valueAsString настолько проста, что достаточно в ObjC иметь простую именованную функцию типа SpeedToString

AntonM030481 replied:
Хотелось универсальное решение втч и для расстояния, когда в зависимости от длины используются крупные или мелкие юниты. А для этого логично получать пару размера и юнитов одновременно.
А что именно с классом не нравится? Вызов 3 C++ функций вместо одной?

Кстати, сейчас есть такое, например:
https://github.com/organicmaps/organicmaps/blob/c2c1f979ac9d9422eae5ae72dad1975b49a47061/iphone/Maps/UI/PlacePage/Components/ElevationProfile/ElevationProfilePresenter.swift#L243-L244

AntonM030481 replied:
Я бы предпочел просто использовать статические переменные вместо того чтобы засорять классы ненужными переменными. По тестам быстродействие вполне нормальное.
https://github.com/AntonM030481/organicmaps/blob/6d5d5ad891141932f9263f66556e9e052ec55f58/platform/localization.cpp#L19-L45

vng commented on iphone/Maps/UI/CarPlay/CarPlayMapViewController.swift:
```diff
@@ -38,73 +40,117 @@ final class CarPlayMapViewController: MWMViewController {
+    } else { // !isCameraOnRoute
+      BlinkSpeedCamLimit(blink: false)
       currentSpeedLabel.textColor = UIColor.speedLimitDarkGray()
+      if let speedLimitMps = speedLimitMps {
```
Я наверное понимаю как это, но вызывает улыбку :) 
if speedLimitMps { не одно и то же?

AntonM030481 replied:
У нас в коде такой подход постоянно встречается:
https://stackoverflow.com/questions/29322977/whats-the-difference-between-if-nil-optional-and-if-let-optional
Временная переменная вместо сравнения с nil чтобы потом не делать unwrap из optional.

biodranik replied:
Вопросов нет, если это optional. Вопросы есть, если это обычная переменная )

biodranik replied:
А, вижу, там optional. Странно конечно, можно было бы сделать оптимальнее.

Вообще, старый swift код был написан до нас, без нашего ревью. Поэтому слепо применять старые практики не стоит. Код должен быть читабельный, простой, не вызывать вопросов, и не иметь ненужного оверхеда в рантайме.

AntonM030481 replied:
добавить self?

vng replied:
Смотри сам, особенно если в swift так принято. Мне нормально по всякому

AntonM030481 replied:
Я не знаю как принято. Написал свою первую строчку на Swift неделю назад)
Использовать то же самое имя при развертывании - логично.
Нужно ли использовать self - наверное, зависит от контекста.
Не нашел с ходу ничего в coding style, кроме того что сравниваем с nil, если развернутое значение не нужно.

AntonM030481 replied:
пока оставлю как есть.

vng commented on platform/localization.hpp:
```diff
@@ -18,5 +23,7 @@ extern std::string GetLocalizedMyPositionBookmarkName();
 extern LocalizedUnits GetLocalizedDistanceUnits();
 extern LocalizedUnits GetLocalizedAltitudeUnits();
+
+extern std::string GetLocalizedSpeedUnits(measurement_utils::Units units);
```
Хм, действительно ли нам нужно показывать это наружу? Почему недостаточно GetLocalizedSpeedUnits() текущей? Иначе зачем клиентскому коду тоже вызывать Settings::Get ...

AntonM030481 replied:
Идея была в том, чтобы не вызывать лишний раз 
`settings::TryGet(settings::kMeasurementUnits, units);`
Сейчас между localisation и measurement_utils много ненужных связей.
И по-хорошему это надо переделать.
На этот раз я решил предоставить унифицированный интерфейс для ios (конечный клиент ничего не знает о системе измерений, все локализовано в одном Core классе).
А в следующий раз наберусь сил и доделаю переделки, описанные выше.

vng commented on platform/measurement_utils.hpp:
```diff
@@ -32,6 +32,7 @@ inline double NauticalMilesToMeters(double nmi) { return nmi * 1852; }
 inline double KmphToMps(double kmph) { return kmph * 1000 / 3600; }
 
 double ToSpeedKmPH(double speed, measurement_utils::Units units);
+double MpsToUnits(double mps, measurement_utils::Units units);
```
Мы же тут уже внутри namespace measurement_utils

AntonM030481 replied:
OK

vng commented on platform/measurement_utils.cpp:
```diff
@@ -224,14 +224,19 @@ string FormatSpeed(double metersPerSecond)
 }
 
-string FormatSpeedNumeric(double metersPerSecond, Units units)
+double MpsToUnits(double metersPerSecond, measurement_utils::Units units)
```
Мы же тут уже внутри namespace measurement_utils

AntonM030481 replied:
OK

AntonM030481 commented on android/jni/com/mapswithme/maps/Framework.cpp:
```diff
@@ -1202,7 +1202,7 @@ Java_com_mapswithme_maps_Framework_nativeGetRouteFollowingInfo(JNIEnv * env, jcl
 
   auto const & rm = frm()->GetRoutingManager();
-  auto const isSpeedLimitExceeded = rm.IsRoutingActive() ? rm.IsSpeedLimitExceeded() : false;
+  auto const isSpeedLimitExceeded = rm.IsRoutingActive() ? rm.IsSpeedCamLimitExceeded() : false;
```
OK

AntonM030481 commented on iphone/Maps/Classes/CustomViews/NavigationDashboard/Views/NavigationControlView.swift:
```diff
@@ -155,23 +155,34 @@ final class NavigationControlView: SolidTouchView, MWMTextToSpeechObserver, MapO
-    if (info.speedLimit != "") {
-      speed += " / " + info.speedLimit;
+    var speedMps = 0.0
+    if let s = LocationManager.lastLocation()?.speed, s > 0 {
```
https://developer.apple.com/documentation/corelocation/cllocation/1423798-speed
A negative value indicates an invalid speed.

AntonM030481 commented on iphone/Maps/Classes/CustomViews/NavigationDashboard/Views/NavigationControlView.swift:
```diff
@@ -155,23 +155,34 @@ final class NavigationControlView: SolidTouchView, MWMTextToSpeechObserver, MapO
+      speedLabel.textColor = UIColor.white()
+      speedBackground.backgroundColor = UIColor.buttonRed()
+    }
+    else {
```
OK

AntonM030481 commented on iphone/Maps/UI/CarPlay/CarPlayMapViewController.swift:
```diff
@@ -38,73 +40,117 @@ final class CarPlayMapViewController: MWMViewController {
+        isSpeedCamBlinking = true
+      }
+    }
+    else {
```
OK

AntonM030481 commented on iphone/Maps/Classes/CustomViews/NavigationDashboard/MWMNavigationDashboardEntity.h:
```diff
@@ -2,25 +2,24 @@
-@property(copy, nonatomic, readonly) NSArray<MWMRouterTransitStepInfo *> * transitSteps;
-@property(copy, nonatomic, readonly) NSAttributedString * estimate;
+@property(copy, nonatomic) NSArray<MWMRouterTransitStepInfo *> * transitSteps;
+@property(copy, nonatomic) NSAttributedString * estimate;
```
OK

AntonM030481 replied:
Вернул readonly в публичном интерфейсе [MWMNavigationDashboardEntity.h]
Вернул передекларирование без readonly в [MWMNavigationDashboardManager+Entity.mm]
для возможности инициализации.
Но этого оказалось недостаточно. Происходит падение при первом же присваивании.
Помог возврат к передекларированию без readonly в [MWMNavigationDashboardEntity.mm]. Но почему - непонятно. В этом mm нет присваивания этим полям и все компилируется.

AntonM030481 replied:
@biodranik как лучше поступить в такой ситуации?

biodranik replied:
Конечно будет падать, если в публичном интерфейсе read only. А как раньше работало? Раньше публично никто не менял?

AntonM030481 replied:
В публичном readonly.
В  [MWMNavigationDashboardManager+Entity.mm] переопределяется как writable и там же меняется.
Я не понимаю, почему приходится переопределять в [MWMNavigationDashboardEntity.mm], поскольку там не вносятся изменения.

biodranik replied:
Т.е. вылетает в текущей версии кода в PR?

AntonM030481 replied:
Удалил MWMNavigationDashboardEntity.mm

biodranik commented:
> Я просто все заново красиво форс репушил, и PR закрылся в процессе (откатился на самое начало). А кнопку Reopen я не заметил.

Форс репуш никак не должен закрывать PR. Это было что-то другое.

biodranik commented:
@AntonM030481 нужно отредактировать коммит и убрать правки сабмодулей.

AntonM030481 commented:
Убраны правки сабмодулей.
Знать бы еще откуда они берутся...
Я их руками не правлю, и через git их не меняю.

AntonM030481 commented:
```
(base) antonmakouski@x86_64-apple-darwin13 organicmaps % git status      
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   3party/boost (new commits)
        modified:   3party/icu/icu (new commits, modified content, untracked content)
        modified:   tools/kothic (new commits)

no changes added to commit (use "git add" and/or "git commit -a")
```

vng commented:
New commits??
git submodule update --init --recursive

AntonM030481 commented:
Submodule path '3party/boost/libs/utility': checked out 'fe417f623790318068b8de8eec0d0cb963e9cb44'
Submodule path '3party/boost/more': checked out '631827729ba7dbed212697ab0f6d61a0748d00a5'
Submodule path '3party/boost/tools/boost_install': checked out '85e1406a570cb8aa13c8b1f885f7cc2a9982c783'
From https://github.com/boostorg/boostbook
 * [new tag]         boost-1.80.0.beta1 -> boost-1.80.0.beta1
Submodule path '3party/boost/tools/boostbook': checked out 'a6701f5d278daba797697a1e5000ee41f25bcaff'
error: The following untracked working tree files would be overwritten by checkout:
        icu4c/source/data/unidata/ucdterms.txt
Please move or remove them before you switch branches.
Aborting
From https://github.com/organicmaps/kothic
 * branch            d6a507cfc09638d489588f2cbd32569f34004318 -> FETCH_HEAD
Submodule path 'tools/kothic': checked out 'd6a507cfc09638d489588f2cbd32569f34004318'
Unable to checkout '904cf62457de2440e8526deb75c95a3f7296f517' in submodule path '3party/icu/icu'
(base) antonmakouski@x86_64-apple-darwin13 organicmaps % rm icu4c/source/data/unidata/ucdterms.txt

AntonM030481 commented:
Локальную проблему с сабмодулями решил зайдя в каждый и вручную сделав чекаут на ревизии указанные [тут](https://github.com/organicmaps/organicmaps/tree/master/3party)

AntonM030481 commented:
Почему-то тайм-аут для map_tests.
У меня локально все нормально проходит.

AntonM030481 commented:
В этом PR я далеко не закончил рефакторингом.
Там еще очень много предстоит, но ИМХО стало лучше, так что предлагаю мержить.

vng reviewed (approved):
@biodranik Давайте уже помержим
